Added support for sub-directory installed check comparison using semantic versioning

icnocop opened the pull request:
Fixes #153

dblock reviewed (commented):
I took a cursory look.

Shouldn't semver comparison work and be documented as something that works for all the places where version comparison worked before? Meaning why is this just for directories?

icnocop commented:
> Shouldn't semver comparison work and be documented as something that works for all the places where version comparison worked before? Meaning why is this just for directories?

I think support for semver comparison can be added to `check_registry_value` and `check_product`, but not for `check_file` because `check_file` calls the Win32 API `GetFileVersionInfo` which returns a `VS_FIXEDFILEINFO` structure which stores version number parts as `DWORD` values.

I think support for semver comparison with other types of installed checks can be added later, if/when needed.

dblock commented:
> > Shouldn't semver comparison work and be documented as something that works for all the places where version comparison worked before? Meaning why is this just for directories?
> 
> I think support for semver comparison can be added to `check_registry_value` and `check_product`, but not for `check_file` because `check_file` calls the Win32 API `GetFileVersionInfo` which returns a `VS_FIXEDFILEINFO` structure which stores version number parts as `DWORD` values.
> 
> I think support for semver comparison with other types of installed checks can be added later, if/when needed.

Fair enough. Ensure that errors are thrown in all the impossible sitiuations.

icnocop commented:
> Ensure that errors are thrown in all the impossible sitiuations.

I've just added unit tests for expected exceptions.

> I see you sneaked in some logger manager ...

Right; I wanted to retain the logging messages that were previously in dotNetInstallerLib\InstalledCheckProduct.cpp.
Some of that code was refactored and moved to dotNetInstallerToolsLib\VersionUtil.cpp.

Code in dotNetInstallerLib uses a macro to log messages which calls a shared static InstallerLog instance.

I figured using a logger manager would be the best option to allow logging messages from dotNetInstallerToolsLib.

Do you have any other recommendations?

Thank you.

dblock commented:
Merge at will.
